---
title: "Hepatocarcinoma Multi-Omics Differential Analysis"
author: "Robert M Flight"
date: last-modified
date-format: YYYY-MM-DD HH:mm
format: 
  html:
    embed-resources: true
    toc: true
---

```{r}
#| label: setup
#| include: false
knitr::opts_chunk$set(echo = FALSE,
											warning = FALSE,
											message = FALSE)
## target knits qmds in their own session, so load libraries here.
targets::tar_source("./packages.R")
tar_source("./R")
```

```{r}
#| label: load-targets
#| include: false
tar_load(c(rna_de_treatment,
					 rna_de_patient,
					 novagene_rna_stats,
					 metabolomics_de_compare,
					 metabolomics_de_aov_compare,
					 metabolomics_treatment_enrichment_reactome,
					 metabolomics_patient_enrichment_reactome,
					 rna_treatment_enrichment_reactome,
					 rna_treatment_enrichment_go,
					 rna_patient_enrichment_reactome,
					 rna_treatment_enrichment_kegg,
					 metabolomics_treatment_enrichment_kegg,
					 rna_patient_enrichment_grouped_go,
					 rna_treatment_enrichment_grouped_go,
					 rna_patient_enrichment_grouped_eachgo,
					 rna_patient_enrichment_grouped_reactome,
					 rna_patient_enrichment_grouped_eachreactome,
					 rna_treatment_enrichment_grouped_reactome))

```

## Purpose

Differential analyses of transcriptomics and metabolomics (bioamines, primary metabolism, lipidomics) for several patients with hepatocarcinoma.

There is an [Executive Summar](#executive-summary) at the end of this document.

## Data

Transcriptomics and metabolomics datasets from hepatocarcinoma.
Transcriptomics samples are normalized and statistical results reported by {DESeq2}.
Each metabolomics type were analyzed separately.
Metabolomics samples are normalized and statistical results reported by {DESeq2}, using a `local` fit for the modeling of dispersions of variance to mean abundance.

For all datasets, we merged the two replicate samples normal adjacent samples from Patient 6, either by adding the abundances via {DESeq2}.

For each dataset, we performed the differential calculations two ways:

  1. Ignoring the paired nature of the samples (treatment).
  1. Incorporating pairing information of the patients (patient).
  
Gene ontology (GO) and Reactome pathway enrichments were run on the transcriptomics differentially expressed genes (p-adjust <= 0.01).

## P-Value Checks

### Transcriptomics

In addition to our own differential results, we also have the values calculated by NovaGene.
We will calculate ratios of $-1 \times log_{10}(p-values)$ just to see which of the methods (NovaGene, treatment, patient) seems to be performing better.

```{r}
#| label: transcriptomics-comparisons
patient_treatment = dplyr::left_join(rna_de_patient[, c("feature_id", "pvalue", "padj")],
																		 rna_de_treatment[, c("feature_id", "pvalue", "padj")],
																		 by = "feature_id", suffix = c(".patient", ".treatment"))
patient_treatment = dplyr::left_join(patient_treatment, novagene_rna_stats[, c("gene_id", "pvalue", "padj")], 
																		 by = c("feature_id" = "gene_id"),
																		 suffix = c(".patient", ".novagene"))
patient_treatment = patient_treatment |>
	dplyr::mutate(pvalue.novagene = pvalue,
								padj.novagene = padj) |>
	dplyr::mutate(patient_treatment = -1*log10(pvalue.patient) - -1*log10(pvalue.treatment),
								treatment_novagene = -1*log10(pvalue.treatment) - -1*log10(pvalue.novagene),
								patient_novagene = -1*log10(pvalue.patient) - -1*log10(pvalue.novagene))
```

The p-value histograms imply that there is definitely a biological signal in the comparison of non-adjacent normal to cancer.
@fig-transcriptomics-pvalues shows the histograms of p-values from each method.
Surprisingly, the number of very low NovaGene p-values seem to be in-between the treatment and patient values.


```{r}
#| label: fig-transcriptomics-pvalues
#| fig-cap: P-Value histogram for each of the different sets of p-values.
patient_treatment_pvals = patient_treatment |>
	dplyr::select(feature_id, pvalue.patient, pvalue.treatment, pvalue.novagene) |>
	tidyr::pivot_longer(-feature_id, names_to = "type", values_to = "pvalue") |>
	dplyr::mutate(type = gsub("pvalue.", "", type)) |>
	dplyr::mutate(type = factor(type, levels = c("novagene", "treatment", "patient"), ordered = TRUE))
patient_treatment_pvals |>
	ggplot(aes(x = pvalue)) +
	geom_histogram(binwidth = 0.01) +
	facet_wrap(~ type, ncol = 1)
```

@fig-transcriptomics-comparison shows that in general, it appears that using the paired patient information results in more extreme p-values than not, which is expected.
In addition, when considering the number of significant genes (those with adjusted p-values <= 0.05), there is definitely a much larger number from the patient analyzed data, as shown in @fig-transcriptomics-upset.

```{r}
#| label: fig-transcriptomics-comparison
#| fig-cap: Sina plot of comparisons of $-1 \times log_{10}(p-values)$ from differential calculations between treatment vs NovaGene; patient vs treatment; and patient vs NovaGene. Red line denotes no difference.
patient_treatment_diff = patient_treatment |>
	dplyr::select(feature_id, patient_treatment, treatment_novagene, patient_novagene) |>
	tidyr::pivot_longer(-feature_id,
											names_to = "comparison",
											values_to = "difference") |>
	dplyr::mutate(comparison = factor(comparison, levels = c("patient_novagene", "treatment_novagene", "patient_treatment"), ordered = TRUE))
patient_treatment_diff |>
	dplyr::filter(dplyr::between(difference, -2.5, 2.5)) |>
	ggplot(aes(x = difference, y = comparison)) +
	geom_sina(alpha = 0.5, size = 0.5) +
	geom_vline(xintercept = 0, color = "red")
```


```{r}
#| label: fig-transcriptomics-upset
#| fig-cap: UpSet plot of significant features from transcriptomics from each method.
transcriptomics_sig = list(patient = patient_treatment |>
													 	dplyr::filter(padj.patient <= 0.05) |>
													 	dplyr::pull(feature_id),
													 treatment = patient_treatment |>
													 	dplyr::filter(padj.treatment <= 0.05) |>
													 	dplyr::pull(feature_id),
													 novagene = patient_treatment |>
													 	dplyr::filter(padj.novagene <= 0.05) |>
													 	dplyr::pull(feature_id))
ComplexHeatmap::UpSet(ComplexHeatmap::make_comb_mat(transcriptomics_sig))
```


So, it makes sense to only look at the results from incorporating the patients in the transcriptomics.

### Metabolomics

The p-value histograms for the metabolomics datasets in @fig-metabolomics-comparisons also show definite biological differences.
@fig-metabolomics-sina also shows that the p-values considering patient are better than just consideration of treatment, at least for the bioamines.
@tbl-metabolomics-diffs counts how many metabolites have better or worse p-values with or without consideration of the patient.
There are a larger number of worse p-values than better p-values when patient is included, for lipidomics and primary-metabolism, which is **not** expected.
We are not sure why this is.
It happens when using t-test statistics, ANOVA, and {DESeq2}.

However, there are more things specifically significant in the patient considered results for both lipidomics (@fig-upset-lipidomics) and primary metabolism (@fig-upset-pm).

```{r}
#| label: fig-metabolomics-comparisons
#| fig-cap: P-Value histograms for treatment and patient differential analyses across each metabolomics platform using t-test.
metabolomics_de_compare = metabolomics_de_compare |>
	dplyr::mutate(type = dplyr::case_when(
		grepl("bioamine", feature_id) ~ "bioamines",
		grepl("pm$", feature_id) ~ "pm",
		grepl("lipids", feature_id) ~ "lipidomics"
	))
metabolomics_pvalue_long = metabolomics_de_compare |>
	dplyr::select(feature_id, pvalue.patient, pvalue.treatment, type) |>
	tidyr::pivot_longer(c(-feature_id, -type), names_to = "pvalue_type",
											values_to = "p.value") |>
	dplyr::mutate(pvalue_type = gsub("pvalue.", "", pvalue_type))
metabolomics_pvalue_long |>
	ggplot(aes(x = p.value)) +
	geom_histogram(binwidth = 0.05) +
	facet_grid(pvalue_type ~ type)
```


```{r}
#| label: fig-metabolomics-sina
#| fig-cap: Sina plot of comparisons of $-1 \times log_{10}(p-values)$ from differential calculations between patient vs treatment for all three metabolomics types using a t-test.
metabolomics_diffs = metabolomics_de_compare |>
	dplyr::mutate(patient_treatment = -1*log10(pvalue.patient) - -1*log10(pvalue.treatment)) |>
	dplyr::select(type, feature_id, patient_treatment) |>
	tidyr::pivot_longer(c(-type, -feature_id),
											names_to = "comparison",
											values_to = "difference")
metabolomics_diffs |>
	ggplot(aes(x = difference, y = comparison)) +
	geom_sina(size = 0.1) +
	geom_vline(xintercept = 0, color = "red") +
	facet_wrap(~ type, nrow = 1)
```

```{r}
#| label: tbl-metabolomics-diffs
#| tbl-cap: Number of metabolites with better or worse p-values when including patient information using t-test.
metabolomics_diff_summary = metabolomics_diffs |>
	dplyr::group_by(type) |>
	dplyr::summarise(`Patient Worse` = sum(difference <= 0),
									 `Patient Better` = sum(difference > 0))
gt::gt(metabolomics_diff_summary)
```

```{r}
#| label: fig-upset-bioamines
#| fig-cap: UpSet plot of significant features for bioamines, considering and ignoring patient.
diff_split_type = split(metabolomics_de_compare, metabolomics_de_compare$type)
bioamines_compare = list(patient = diff_split_type$bioamines |>
													 	dplyr::filter(padj.patient <= 0.05) |>
													 	dplyr::pull(feature_id),
													 treatment = diff_split_type$bioamines |>
													 	dplyr::filter(padj.treatment <= 0.05) |>
													 	dplyr::pull(feature_id))
ComplexHeatmap::UpSet(ComplexHeatmap::make_comb_mat(bioamines_compare), column_title = "Bioamines")
```

```{r}
#| label: fig-upset-lipidomics
#| fig-cap: UpSet plot of significant features for lipidomics, considering and ignoring patient.
lipidomics_compare = list(patient = diff_split_type$lipidomics |>
													 	dplyr::filter(padj.patient <= 0.05) |>
													 	dplyr::pull(feature_id),
													 treatment = diff_split_type$lipidomics |>
													 	dplyr::filter(padj.treatment <= 0.05) |>
													 	dplyr::pull(feature_id))
ComplexHeatmap::UpSet(ComplexHeatmap::make_comb_mat(lipidomics_compare), column_title = "Lipidomics")
```

```{r}
#| label: fig-upset-pm
#| fig-cap: UpSet plot of significant features for primary metabolism, considering and ignoring patient.	
pm_compare = list(patient = diff_split_type$pm |>
													 	dplyr::filter(padj.patient <= 0.05) |>
													 	dplyr::pull(feature_id),
													 treatment = diff_split_type$pm |>
													 	dplyr::filter(padj.treatment <= 0.05) |>
													 	dplyr::pull(feature_id))
ComplexHeatmap::UpSet(ComplexHeatmap::make_comb_mat(pm_compare), column_title = "Primary Metabolism")
```

## Enrichment Results

### Comparing -Omics

One way to try and integrate the results from the metabolomics and transcriptomics, is to look for pathway enrichment intersections from both the metabolomics and transcriptomics.
Those results are shown in @fig-enrichment-intersections-reactome for Reactome and @fig-enrichment-intersections-kegg for KEGG.
Zip, zero, nothing in the case of Reactome and 1 in the case of KEGG, and it does not seem to be very informative.

```{r}
#| label: fig-enrichment-intersections-reactome
#| fig-cap: UpSet plot comparing the various enrichment results from Reactome pathways.
list_enrich = list(rna_patient_neg = rna_patient_enrichment_reactome$stats$neg |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 rna_patient_pos = rna_patient_enrichment_reactome$stats$pos |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 rna_patient_all = rna_patient_enrichment_reactome$stats$all |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 rna_treatment_neg = rna_treatment_enrichment_reactome$stats$neg |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 rna_treatment_pos = rna_treatment_enrichment_reactome$stats$pos |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 rna_treatment_all = rna_treatment_enrichment_reactome$stats$all |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 metabolomics_treatment_neg = metabolomics_treatment_enrichment_reactome$stats$neg |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 metabolomics_treatment_all = metabolomics_treatment_enrichment_reactome$stats$all |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 metabolomics_patient_neg = metabolomics_patient_enrichment_reactome$stats$neg |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID),
									 metabolomics_patient_all = metabolomics_patient_enrichment_reactome$stats$all |>
									 	dplyr::filter(padjust <= 0.05) |>
									 	dplyr::pull(ID))

ComplexHeatmap::UpSet(ComplexHeatmap::make_comb_mat(list_enrich))
```

```{r}
#| label: fig-enrichment-intersections-kegg
#| fig-cap: UpSet plot comparing the various enrichment results from KEGG pathways.
kegg_enrich = list(rna_treatment_neg = rna_treatment_enrichment_kegg$stats$neg |>
									 	dplyr::filter(p <= 0.05) |>
									 	dplyr::pull(ID),
									 rna_treatment_pos = rna_treatment_enrichment_kegg$stats$pos |>
									 	dplyr::filter(p <= 0.05) |>
									 	dplyr::pull(ID),
									 rna_treatment_all = rna_treatment_enrichment_kegg$stats$all |>
									 	dplyr::filter(p <= 0.05) |>
									 	dplyr::pull(ID),
									 metabolomics_treatment_neg = metabolomics_treatment_enrichment_kegg$stats$neg |>
									 	dplyr::filter(p <= 0.05) |>
									 	dplyr::pull(ID),
									 metabolomics_treatment_all = metabolomics_treatment_enrichment_kegg$stats$all |>
									 	dplyr::filter(p <= 0.05) |>
									 	dplyr::pull(ID))

ComplexHeatmap::UpSet(ComplexHeatmap::make_comb_mat(kegg_enrich))
```

```{r}
#| label: tbl-kegg-intersect
#| tbl-cap: KEGG pathways that are enriched in both the metabolomics and RNA datasets.
kegg_match = base::intersect(kegg_enrich$rna_treatment_neg, kegg_enrich$metabolomics_treatment_neg)
rna_kegg_table = rna_treatment_enrichment_kegg$stats$neg |>
	dplyr::filter(ID %in% kegg_match)
gt::gt(rna_kegg_table |> dplyr::select(ID, description))
```

### Transcriptomics Enrichments

For the transcriptomics experiment, we did Gene Ontology (GO) and Reactome pathway enrichment for **all** differentially expressed genes (@tbl-rna-patient-go-all, @tbl-rna-patient-reactome-all), as well as the **negatively** (@tbl-rna-patient-go-neg, @tbl-rna-patient-reactome-neg) and **positively** (@tbl-rna-patient-go-pos, @tbl-rna-patient-reactome-pos) fold-change differentially expressed genes.

Columns are:

  * p: p-value;
  * odds: odds ratio, roughly how many were observed vs expected;
  * expected: how many genes expected based on the size of the differential list and the universe of measured genes;
  * counts: how many genes were observed;
  * padjust: Benjamini-Hochberg adjusted p-value;
  * ID: Gene Ontology or Reactome pathway ID;
  * description: text description;
  * group: group designation, denotes highly related annotations by shared annotated genes.

Annotations are grouped based on the common genes between annotations.
A similarity cutoff of 0.8 was used before determining communities of annotations.

Full tables should have been provided in an Excel file.

```{r}
#| label: tbl-rna-patient-go-all
#| tbl-cap: Grouped GO terms for transcriptomics paired patient sample differential analysis, irrespective of fold-change direction.
go_all = rna_patient_enrichment_grouped_eachgo$all
go_print_all = go_all |>
	dplyr::filter(grepl("MF:lipoprotein particle binding|CC:vesicle lumen|BP:regulation of tube size|BP:regulation of lipase activity|lipo.*|CC:protein-lipid complex|BP:regulation of spindle checkpoint|BP:regulation of chromosome segregation|BP:negative regulation of cell cycle", group)) |>
	dplyr::arrange(group, padjust)
weird_row = go_print_all$ID %in% "GO:0016098"
go_all_out = gt::gt(go_print_all) |>
	gt::fmt_number(columns = ends_with(c("odds", "expected", "counts"))) |>
	gt::fmt_scientific(columns = ends_with(c("p", "padjust"))) |>
	gt::cols_width(ends_with(c("p", "padjust")) ~ px(400))
go_all_out
```

```{r}
#| label: tbl-rna-patient-go-neg
#| tbl-cap: Grouped GO terms for transcriptomics paired patient sample differential analysis, negative fold-change direction.
go_neg = rna_patient_enrichment_grouped_eachgo$neg
go_print_neg = go_neg |>
	dplyr::filter(grepl("MF:lipoprotein particle binding|CC:vesicle lumen|BP:regulation of tube size|BP:regulation of lipase activity|lipo.*|CC:protein-lipid complex|BP:regulation of spindle checkpoint|BP:regulation of chromosome segregation|BP:negative regulation of cell cycle", group)) |>
	dplyr::arrange(group, padjust)
weird_row = go_print_neg$ID %in% "GO:0016098"
go_neg_out = gt::gt(go_print_all) |>
	gt::fmt_number(columns = ends_with(c("odds", "expected", "counts"))) |>
	gt::fmt_scientific(columns = ends_with(c("p", "padjust"))) |>
	gt::cols_width(ends_with(c("p", "padjust")) ~ px(400))
go_neg_out
```

```{r}
#| label: tbl-rna-patient-go-pos
#| tbl-cap: Grouped GO terms for transcriptomics paired patient sample differential analysis, positive fold-change direction.
go_pos = rna_patient_enrichment_grouped_eachgo$pos
go_print_pos = go_pos |>
	dplyr::filter(grepl("MF:lipoprotein particle binding|CC:vesicle lumen|BP:regulation of tube size|BP:regulation of lipase activity|lipo.*|CC:protein-lipid complex|BP:regulation of spindle checkpoint|BP:regulation of chromosome segregation|BP:negative regulation of cell cycle", group)) |>
	dplyr::arrange(group, padjust)
weird_row = go_print_pos$ID %in% "GO:0016098"
go_pos_out = gt::gt(go_print_pos) |>
	gt::fmt_number(columns = ends_with(c("odds", "expected", "counts"))) |>
	gt::fmt_scientific(columns = ends_with(c("p", "padjust"))) |>
	gt::cols_width(ends_with(c("p", "padjust")) ~ px(400))
go_pos_out
```

```{r}
#| label: tbl-rna-patient-reactome-all
#| tbl-cap: Grouped Reactome pathways for transcriptomics paired patient sample differential analysis, for all differential genes.
reactome_all = rna_patient_enrichment_grouped_eachreactome$all |>
	dplyr::filter(!group %in% "") |>
	dplyr::arrange(group, padjust)
reactome_print_all = gt::gt(reactome_all) |>
	gt::fmt_number(columns = ends_with(c("odds", "expected", "counts"))) |>
	gt::fmt_scientific(columns = ends_with(c("p", "padjust"))) |>
	gt::cols_width(ends_with(c("p", "padjust")) ~ px(400))
reactome_print_all
```

```{r}
#| label: tbl-rna-patient-reactome-neg
#| tbl-cap: Grouped Reactome pathways for transcriptomics paired patient sample differential analysis, for negatively differential genes.
reactome_neg = rna_patient_enrichment_grouped_eachreactome$all |>
	dplyr::filter(!group %in% "") |>
	dplyr::arrange(group, padjust)
reactome_print_neg = gt::gt(reactome_neg) |>
	gt::fmt_number(columns = ends_with(c("odds", "expected", "counts"))) |>
	gt::fmt_scientific(columns = ends_with(c("p", "padjust"))) |>
	gt::cols_width(ends_with(c("p", "padjust")) ~ px(400))
reactome_print_neg
```

```{r}
#| label: tbl-rna-patient-reactome-pos
#| tbl-cap: Grouped Reactome pathways for transcriptomics paired patient sample differential analysis, for positively differential genes.
reactome_pos = rna_patient_enrichment_grouped_eachreactome$pos |>
	dplyr::filter(!group %in% "") |>
	dplyr::arrange(group, padjust)
reactome_print_pos = gt::gt(reactome_pos) |>
	gt::fmt_number(columns = ends_with(c("odds", "expected", "counts"))) |>
	gt::fmt_scientific(columns = ends_with(c("p", "padjust"))) |>
	gt::cols_width(ends_with(c("p", "padjust")) ~ px(400))
reactome_print_pos
```


## Executive Summary

* Considering patient definitively improves the transcriptomics analysis.
* Considering patient definitively improves the bioamines, and we think it improves the lipidomics and primary metabolism metabolomics.
* Comparing pathway enrichments between transcriptomics and metabolomics doesn't highlight anything interesting.
* Gene ontology enrichment of the transcriptomics results highlight a variety of terms related to cell cycle in the positively differential genes (expected for cancer), as well as some related to lipid processing in the negatively differential genes (expected for this hepatocarcinoma).
* Reactome pathway enrichment highlights many pathways that are likely heavily similar to the GO terms, with the exception of lipid related.
* Full GO and Reactome pathways were provided in an Excel file (`helsley_hcc_transcriptomics_enrichment-2024-04-02.xlsx`).
* Next steps include:
  * Examining what classes and types of metabolites are over-represented in the differential results.
  * Looking for highly correlated metabolites and genes.