---
title: "Hepatocarcinoma Multi-Omics Differential Analysis"
author: "Robert M Flight"
date: last-modified
date-format: YYYY-MM-DD HH:mm
format: 
  html:
    embed-resources: true
    toc: true
---

```{r}
#| label: setup
#| include: false
knitr::opts_chunk$set(echo = FALSE,
											message = FALSE,
											dev = "ragg_png",
											dpi = 300)
targets::tar_source("./packages.R")
tar_source("R")
library(tidygraph)
library(ggraph)
library(dbscan)
```

```{r}
#| label: load-targets
#| include: false
tar_load(c(rna_metabolites_spearman,
					 rna_patient_enrichment_grouped_eachgo,
					 rna_patient_enrichment_go,
					 rna_abundances,
					 rna_de_patient,
					 metabolite_abundances,
					 rna_within_cor,
					 metabolites_within_cor
					 ))
rna_significant = tar_read(rna_de_patient)
metabolites_significant = tar_read(metabolomics_de_patient_list)
```

## Data

Hepatocarcinoma transcriptomics significant entries, and metabolomics significant entries (p-adjusted <= 0.05), Spearman correlated between the genes and metabolites, using the 12 samples that are in common across all 4 datasets.

## Significant Correlations

P-adjusted correlations <= 0.01.
@fig-examine-correlations shows the histogram of significant correlations between genes and metabolites.

@tbl-n-genes-metabolites lists how many metabolites and genes are in the significant correlations.

```{r}
#| label: fig-examine-correlations
#| fig-cap: Number of genes and metabolites having significant correlations.
sig_cor = rna_metabolites_spearman |>
	dplyr::filter(padjust <= 0.01) |>
	dplyr::transmute(gene = s1, metabolite = s2, cor = cor,
									 dir = sign(cor))

rna_sig = rna_significant |>
	dplyr::transmute(gene.lfc = sign(log2FoldChange),
									 gene = feature_id)
met_sig = metabolites_significant |>
	dplyr::transmute(metabolite.lfc = sign(log2FoldChange),
									 metabolite = feature_id)
sig_cor = dplyr::left_join(sig_cor, rna_sig, by = "gene")
sig_cor = dplyr::left_join(sig_cor, met_sig, by = "metabolite")
sig_cor |>
	ggplot(aes(x = cor)) +
	geom_histogram(bins = 100)
```

```{r}
#| label: tbl-n-genes-metabolites
#| tbl-cap: Number of genes and metabolites after filtering to those with significant correlations (p-adjust <= 0.01).
n_each = sig_cor |>
	dplyr::summarise(n_gene = length(unique(gene)),
									 n_metabolite = length(unique(metabolite)))
gt::gt(n_each)
```

```{r}
#| label: fig-n-to-each
#| fig-cap: Histograms of number of metabolites with significant edges to genes (Genes -> Metabolites) and number of genes with significant edges to metabolites (Metabolites -> Genes), split out by direction of correlation and direction of log-fold-change of the first entry.
#| fig-width: 8
#| fig-height: 15
gene_to_metabolite = sig_cor |>
	dplyr::mutate(sign = sign(cor)) |>
	dplyr::group_by(gene, sign, gene.lfc) |>
	dplyr::summarise(g2m = dplyr::n())
metabolite_to_gene = sig_cor |>
	dplyr::mutate(sign = sign(cor)) |>
	dplyr::group_by(metabolite, sign, metabolite.lfc) |>
	dplyr::summarise(m2g = dplyr::n())

gm_plot = gene_to_metabolite |>
	dplyr::mutate(gene.lfc = paste0("lfc ", gene.lfc),
								sign = paste0("cor ", sign)) |>
	ggplot(aes(x = g2m)) +
	geom_histogram(bins = 100) +
	facet_grid(gene.lfc ~ sign) +
	coord_cartesian(ylim = c(0, 250)) +
	labs(subtitle = "Genes -> Metabolites")
mg_plot = metabolite_to_gene |>
	dplyr::mutate(metabolite.lfc = paste0("lfc ", metabolite.lfc),
								sign = paste0("cor ", sign)) |>
	ggplot(aes(x = m2g)) +
	geom_histogram(bins = 100) +
	facet_grid(metabolite.lfc ~ sign) +
	coord_cartesian(ylim = c(0, 200)) +
	labs(subtitle = "Metabolites -> Genes")
gm_plot / mg_plot
```


We can consider this as a simple bipartite graph, and order the nodes by their number of edges and type, as shown in @fig-sig-cor-graph.
We can also ignore the type of node, and just order them by whether the gene or metabolite had positive or negative LFC and the number of edges to the other type of node, see @fig-sig-cor-graph2.
Both of these imply there are definite groups of metabolites all connected to one or more sets of genes.

```{r}
#| label: fig-sig-cor-graph
#| fig-cap: Bipartite graphing of connected genes and metabolites, ordered by node type (gene or metabolite) and number of nodes.
#| fig-width: 9
#| fig-height: 8

all_lfc = dplyr::bind_rows(rna_sig |>
													 	dplyr::transmute(name = gene,
													 									 lfc = gene.lfc),
													 met_sig |>
													 	dplyr::transmute(name = metabolite,
													 									 lfc = metabolite.lfc))

sig_cor_graph = as_tbl_graph(sig_cor, directed = FALSE) |>
	dplyr::mutate(type = dplyr::case_when(
		grepl("^ENS", name) ~ "gene",
		TRUE ~ "metabolite"
	),
	n_connect = tidygraph::centrality_degree())

sig_cor_graph = dplyr::left_join(sig_cor_graph, all_lfc, by = "name")

sig_cor_graph = sig_cor_graph |>
	arrange(type, n_connect)

ggraph(sig_cor_graph, layout = "linear", circular = TRUE) +
	geom_edge_arc(alpha = 0.1) +
	geom_node_point(aes(color = type)) +
	coord_fixed()
```



```{r}
#| label: fig-sig-cor-graph2
#| fig-cap: Ordered by up or down log-fold-change, and number of connections. Color of nodes indicate gene or metabolite, and color of edge is positive or negative correlation.
#| fig-width: 9
#| fig-height: 8
sig_cor_graph = sig_cor_graph |>
	arrange(lfc, n_connect)
ggraph(sig_cor_graph, layout = "linear", circular = TRUE) +
	geom_edge_arc(alpha = 0.1, aes(color = dir)) +
	geom_node_point(size = 0.5, aes(color = type)) +
	coord_fixed()
```

## Check Correlations

We should check what these significant correlations look like in practice, especially compared to gene - gene and metabolite - metabolite correlations.

```{r}
#| label: grab-to-check
set.seed(1134)
check_cor = dplyr::bind_rows(rna_metabolites_spearman |>
														 	dplyr::filter(dplyr::between(cor, -0.5, 0.5)) |>
														 	dplyr::slice_sample(n = 10),
														 rna_metabolites_spearman |>
														 	dplyr::filter(padjust <= 0.01) |>
														 	dplyr::filter(dplyr::between(abs(cor), 0.8, 0.9)) |>
														 	dplyr::slice_sample(n = 10))
check_cor = check_cor |>
	dplyr::arrange(abs(cor))

out_plots = purrr::map(seq_len(nrow(check_cor)), \(in_row){
	use_gene = check_cor$s1[in_row]
	use_met = check_cor$s2[in_row]
	tmp_df = data.frame(gene = rna_abundances[use_gene, ],
											metabolite = metabolite_abundances[use_met, ])
	ggplot(tmp_df, aes(x = gene, y = metabolite)) +
		geom_point() +
		labs(x = use_gene, y = use_met, subtitle = paste0("Spearman: ", format(check_cor$cor[in_row], digits = 2)))
	
})

```

```{r}
#| label: fig-check-cor
#| fig-cap: Gene - metabolite plots for a random selection of non-significant and significant gene - metabolite pairs.
patchwork::wrap_plots(out_plots, nrow = 5, ncol = 4, byrow = TRUE)
```

## Groups of Interesting Genes

We will take some of the interesting groups of gene ontology (GO) terms, and then all the genes annotated to them, verify their correlations among those genes, and then look for significantly correlated metabolites, and hopefully find something interesting.

```{r}
#| label: define-groups
go_neg = rna_patient_enrichment_grouped_eachgo$neg
go_neg_interesting = go_neg |>
	dplyr::filter(grepl("MF:lipoprotein particle binding|CC:vesicle lumen|BP:regulation of tube size|BP:regulation of lipase activity|lipo.*|CC:protein-lipid complex|BP:regulation of spindle checkpoint|BP:regulation of chromosome segregation|BP:negative regulation of cell cycle", group))

go_neg_split = split(go_neg_interesting$ID, go_neg_interesting$group)

rna_neg_sig = rna_de_patient |>
	dplyr::filter(padj <= 0.01, log2FoldChange <= 0) |>
	dplyr::pull(feature_id)
rna_cor_sig = sig_cor |>
	dplyr::pull(gene) |>
	unique()
annotation_list = rna_patient_enrichment_go$enrich@annotation@annotation_features

go_neg_genes = purrr::map(go_neg_split, \(in_go){
	go_genes = base::intersect(unique(unlist(annotation_list[in_go])), rna_neg_sig)
	go_genes = base::intersect(go_genes, rna_cor_sig)
	go_genes_cor = rna_within_cor |>
		dplyr::filter((s1 %in% go_genes) & (s2 %in% go_genes))
	
	gene_metabolite = sig_cor |>
		dplyr::filter(gene %in% go_genes)
	
	metabolite_cor = metabolites_within_cor |>
		dplyr::filter((s1 %in% unique(gene_metabolite$metabolite)) & (s2 %in% unique(gene_metabolite$metabolite))) 
	
	list(genes = go_genes,
			 gene_cor = go_genes_cor,
			 metabolite_cor = metabolite_cor,
			 gene_metabolite = gene_metabolite)
	
})

go_neg_metabolites = purrr::map(go_neg_genes)
```