---
title: "PCA and Differential Heatmaps"
author: "Robert M Flight"
date: last-modified
date-format: YYYY-MM-DD HH:mm
format: 
  docx:
    keep-md: true
  html:
    toc: true
    embed-resources: true
---

```{r}
#| label: setup
#| include: false
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  fig.width = 10,
  fig.height = 5,
  dev = c("ragg_png", "svg"),
  dpi = 600
)
## target knits qmds in their own session, so load libraries here.
targets::tar_source("packages.R")
tar_source("R")
```

```{r}
#| label: load-targets
#| include: false
tar_load(c(
  color_scales,
  bioamines_qcqa,
  lipidomics_qcqa,
  primary_metabolism_qcqa,
  rna_qcqa,
  rna_pca_heatmap,
  bioamines_pca_heatmap,
  lipidomics_pca_heatmap,
  pm_pca_heatmap
))
```

## Methods

Tissue transcript abundances were obtained from RNA-seq data provided by Novogene.
Metabolite abundances for biogenic amines, lipids, and primary metabolism associated metabolites were provided by West Coast Metabolomics Center.
For each -omic modality, values were stored in a `DESeq2` DataSet object, and normalized using `estimateSizeFactors`.
Values less than 10 were set to 0, and only those features with non-missing values in 0.75 of a group of samples were kept (using `visualizationQualityControl::keep_non_missing_percentage`).
Principal components analysis used the `stats::prcomp` function on the normalized values after removing features that were missing in too many samples.
Differential testing was carried out using `DESeq2`, with estimates of the dispersions fit using the parametric method for RNA-seq, and the local method for the metabolomics data sets.
Features with an adjusted p-value <= 0.01 were considered differentially expressed when comparing cancerous to normal adjacent samples.

ANOVA of sample scores on each principal component against **treatment** and **patient** were calculated using `visualizationQualityControl::visqc_test_pca_scores`.
The p-values for PCs 1 - 4 for each of **treatment** and **patient** are reported.
Variances of sample scores on PC1 within treatment were calculated using the median-absolute-deviations (MADs) `stats::mad` function.  

Log2 ratios of cancerous / normal adjacent were calculated for each patient, with infinite, not a number or missing values replaced by missing.
Heatmaps were generated using `ComplexHeatmap` and `scico` colormaps.

```{r}
#| label: refactor-function
refactor_treatment = function(qcqa_obj) {
  # pca_values = rna_qcqa$pca_nooutlier_values
  pca_values = qcqa_obj$pca_nooutlier_values
  pca_values = pca_values |>
    dplyr::mutate(
      Treatment = dplyr::case_when(
        treatment %in% "cancerous" ~ "Cancerous",
        treatment %in% "normal_adjacent" ~ "Normal Adjacent"
      )
    )
  pca_values
}

use_colors = color_scales$normal_cancer
```


## RNA-Seq

```{r}
#| label: fig-rna-seq
#| fig-cap: RNA-seq samples PCA (left), volcano plot (center) and differential gene heatmap (right). Differential genes are colored by the log2-fold-change cancerous / normal adjacent samples in each patient.
rna_pca_heatmap$heatmap = draw(rna_pca_heatmap$heatmap) |> grid.grabExpr()
patchwork::wrap_plots(rna_pca_heatmap, nrow = 1)
```

```{r}
#| label: tbl-rna-seq-anova
#| tbl-cap: Statistics of ANOVA for PCs 1 - 4 vs treatment and patient.
use_pcs = c("PC1", "PC2", "PC3", "PC4")
rna_qcqa$pca_nooutlier_anova |>
  dplyr::filter(PC %in% use_pcs) |>
  dplyr::arrange(dplyr::desc(variable), PC) |>
  dplyr::select(variable, PC, p.value, statistic) |>
  gt::gt() |>
  gt::fmt_engineering(columns = p.value, decimals = 2) |>
  gt::fmt_number(columns = statistic, decimals = 2)
```

```{r}
#| label: tbl-rnaseq-mad
#| tbl-cap: MAD of sample scores on PC1 and number of samples by treatment.
refactor_treatment(rna_qcqa) |>
  dplyr::group_by(Treatment) |>
  dplyr::summarise(MAD = mad(PC1), N = dplyr::n()) |>
  gt::gt() |>
  gt::fmt_number(columns = MAD) |>
  gt::fmt_integer(columns = N)
```


## Lipidomics

```{r}
#| label: fig-lipidomics
#| fig-cap: Lipidomics PCA (left) and differential metabolite heatmap (right). Differential metabolites are colored by the log2-fold-change cancerous / normal adjacent samples in each patient.
lipidomics_pca_heatmap$heatmap = draw(lipidomics_pca_heatmap$heatmap) |>
  grid.grabExpr()

patchwork::wrap_plots(lipidomics_pca_heatmap, nrow = 1)
```

```{r}
#| label: tbl-lipidomics-anova
#| tbl-cap: Statistics of ANOVA for PCs 1 - 4 vs treatment and patient.
lipidomics_qcqa$pca_nooutlier_anova |>
  dplyr::filter(PC %in% use_pcs) |>
  dplyr::arrange(dplyr::desc(variable), PC) |>
  dplyr::select(variable, PC, p.value, statistic) |>
  gt::gt() |>
  gt::fmt_engineering(columns = p.value, decimals = 2) |>
  gt::fmt_number(columns = statistic, decimals = 2)
```

```{r}
#| label: tbl-lipidomics-mad
#| tbl-cap: MAD of sample scores on PC1 and number of samples by treatment.
refactor_treatment(lipidomics_qcqa) |>
  dplyr::group_by(Treatment) |>
  dplyr::summarise(MAD = mad(PC1), N = dplyr::n()) |>
  gt::gt() |>
  gt::fmt_number(columns = MAD) |>
  gt::fmt_integer(columns = N)
```

## Bioamines

```{r}
#| label: fig-bioamines
#| fig-cap: Biogenic amines PCA (left) and differential metabolite heatmap (right). Differential metabolites are colored by the log2-fold-change cancerous / normal adjacent samples in each patient.
bioamines_pca_heatmap$heatmap = draw(bioamines_pca_heatmap$heatmap) |>
  grid.grabExpr()

patchwork::wrap_plots(bioamines_pca_heatmap, nrow = 1)
```

```{r}
#| label: tbl-bioamines-anova
#| tbl-cap: Statistics of ANOVA for PCs 1 - 4 vs treatment and patient.
bioamines_qcqa$pca_nooutlier_anova |>
  dplyr::filter(PC %in% use_pcs) |>
  dplyr::arrange(dplyr::desc(variable), PC) |>
  dplyr::select(variable, PC, p.value, statistic) |>
  gt::gt() |>
  gt::fmt_engineering(columns = p.value, decimals = 2) |>
  gt::fmt_number(columns = statistic, decimals = 2)
```

```{r}
#| label: tbl-bioamines-mad
#| tbl-cap: MAD of sample scores on PC1 and number of samples by treatment.
refactor_treatment(bioamines_qcqa) |>
  dplyr::group_by(Treatment) |>
  dplyr::summarise(MAD = mad(PC1), N = dplyr::n()) |>
  gt::gt() |>
  gt::fmt_number(columns = MAD) |>
  gt::fmt_integer(columns = N)
```

## Primary Metabolism

```{r}
#| label: fig-primary-metabolism
#| fig-cap: Primary metabolism PCA of samples (left) and differential metabolite heatmap (right). Differential metabolites are colored by the log2-fold-change cancerous / normal adjacent samples in each patient.
pm_pca_heatmap$heatmap = draw(pm_pca_heatmap$heatmap) |> grid.grabExpr()

patchwork::wrap_plots(pm_pca_heatmap, nrow = 1)
```

```{r}
#| label: tbl-pm-anova
#| tbl-cap: Statistics of ANOVA for PCs 1 - 4 vs treatment and patient.
primary_metabolism_qcqa$pca_nooutlier_anova |>
  dplyr::filter(PC %in% use_pcs) |>
  dplyr::arrange(dplyr::desc(variable), PC) |>
  dplyr::select(variable, PC, p.value, statistic) |>
  gt::gt() |>
  gt::fmt_engineering(columns = p.value, decimals = 2) |>
  gt::fmt_number(columns = statistic, decimals = 2)
```

```{r}
#| label: tbl-pm-mad
#| tbl-cap: MAD of sample scores on PC1 and number of samples by treatment.
refactor_treatment(primary_metabolism_qcqa) |>
  dplyr::group_by(Treatment) |>
  dplyr::summarise(MAD = mad(PC1), N = dplyr::n()) |>
  gt::gt() |>
  gt::fmt_number(columns = MAD) |>
  gt::fmt_integer(columns = N)
```