---
title: "Liver Carcinoma Omics QC-QA"
author: "Robert M Flight"
date: last-modified
date-format: YYYY-MM-DD HH:mm
format: 
  html:
    embed-resources: true
    toc: true
---

```{r}
#| label: setup
#| include: false
knitr::opts_chunk$set(echo = FALSE)
## target knits qmds in their own session, so load libraries here.
targets::tar_source("./packages.R")
targets::tar_source("R")
```

```{r}
#| label: load-targets
#| include: false
tar_load(c(rna_qcqa,
					 bioamines_qcqa,
					 lipidomics_qcqa,
					 primary_metabolism_qcqa,
					 sample_info,
					 rna_n_qcqa,
					 bioamines_n_qcqa,
					 lipidomics_n_qcqa,
					 pr_n_qcqa,
					 rna_left,
					 bioamines_left,
					 lipidomics_left,
					 pm_left,
					 color_scales))
```

## Purpose

Quality-control and quality-assurance of the various datasets from the Helsley liver carcinoma multi-omics experiment.
An [executive summary](#executive-summary) is available at the end of this document.

## Data

Liver carcinoma and adjacent normal samples from several patients.
RNA-seq transcriptomics performed by Novogene.
Two patients are missing RNA-seq for their normal samples, and a single patient has a technical replicate of the normal sample.
Metabolomics for biogenic amines, lipidomics, and primary metabolism measured by West Coast Metabolomics Center (WCMC).

```{r}
#| label: tbl-number-samples
#| tbl-cap: "Number of samples in each treatment group overall."
tmp_var = as.data.frame(table(sample_info$treatment))
names(tmp_var) = c("treatment", "N-Samples")
gt::gt(tmp_var)
```

## Processing

RNA-seq data was normalized by {DESeq2}.
Metabolomics datasets normalized by dividing by the median of all values in each sample.
For each dataset, we only keep those features (genes or metabolites) that were present in at least 75% of the normal or cancer samples.
Sample - sample correlations were calculated using information-content-informed Kendall-tau (ICI-Kt).
Principal component analysis (PCA) was performed after $log(x + 1)$ (using the {log1p} function).

In the heatmaps, the sample-sample correlations are mapped to colors.
The values are reflected across the diagonal (so top and bottom are equivalent), and the diagonal shows the fraction of non-missing entries for that sample.

In the median correlation plots, the samples are grouped by treatment, and the median correlation values **within** the treatment group are calculated for each sample in that treatment group.
**Sina** plots attempt to plot the distribution of values within each grouping so that the individual values are visible.

For PCA, each principal component (PC) is a linear combination of all of the feature abundances, ranked by how much variance is accounted for in that combination, and where each subsequent PC is orthogonal to the previous one.
The PC's are labeled with the percentage of variance that PC accounts for out of all of the PCs.


## RNA Seq

Number of features:

  * Total features: `r rna_n_qcqa[1]`
  * Features in 75% of a class: `r rna_n_qcqa[2]`
  
```{r}
#| label: fig-rnaseq-heatmap
#| fig-cap: Sample-sample ICI-Kt correlations for RNA-seq data.
rna_qcqa$heatmap
```

```{r}
#| label: fig-rnaseq-mediancor
#| fig-cap: Median ICI-Kt correlations for RNA-seq data.
rna_qcqa$median_cor |>
		ggplot(aes(x = treatment, y = med_cor, color = treatment, shape = outlier, group = treatment)) +
		geom_sina() +
		scale_color_manual(values = color_scales$treatment) +
	labs(y = "Median Correlation")
```

```{r}
#| label: fig-rnaseq-pca
#| fig-cap: PCA plot for RNA-seq data.
rna_qcqa$pca_all
```


## Primary Metabolism

Number of features:

  * Total: `r pr_n_qcqa[1]`
  * In 75% of a class: `r pr_n_qcqa[2]`

```{r}
#| label: fig-pr-heatmap
#| fig-cap: Sample-sample ICI-Kt correlations for primary metabolism.
primary_metabolism_qcqa$heatmap
```

```{r}
#| label: fig-pr-mediancor
#| fig-cap: Median ICI-Kt correlations for primary metabolism data.
primary_metabolism_qcqa$median_cor |>
		ggplot(aes(x = treatment, y = med_cor, color = treatment, shape = outlier, group = treatment)) +
		geom_sina() +
		scale_color_manual(values = color_scales$treatment) +
	labs(y = "Median Correlation")
```

```{r}
#| label: fig-pr-pca-all
#| fig-cap: PCA plot for primary metabolism data, including pooled samples.
primary_metabolism_qcqa$pca_all
```

```{r}
#| label: fig-pr-pca-nopool
#| fig-cap: PCA plot for primary metabolism data, no pooled samples.
primary_metabolism_qcqa$pca_noblanks
```

```{r}
#| label: fig-pr-pca-nooutlier
#| fig-cap: PCA plot for primary metabolism metabolites, no outliers.
primary_metabolism_qcqa$pca_nooutlier
```


## Biogenic Amines

Number of features:

  * Total: `r bioamines_n_qcqa[1]`
  * In 75% of a class: `r bioamines_n_qcqa[2]`

```{r}
#| label: fig-bioamines-heatmap
#| fig-cap: Sample-sample ICI-Kt correlations for biogenic amines metabolites.
bioamines_qcqa$heatmap
```

```{r}
#| label: fig-bioamines-mediancor
#| fig-cap: Median ICI-Kt correlations for biogenic amines metabolites.
bioamines_qcqa$median_cor |>
		ggplot(aes(x = treatment, y = med_cor, color = treatment, shape = outlier, group = treatment)) +
		geom_sina() +
		scale_color_manual(values = color_scales$treatment) +
	labs(y = "Median Correlation")
```

```{r}
#| label: fig-bioamines-pca
#| fig-cap: PCA plot for biogenic amines metabolites.
bioamines_qcqa$pca_all
```

```{r}
#| label: fig-bioamines-pca-nopool
#| fig-cap: PCA plot for biogenic amines metabolites, without pooled and blank samples.
bioamines_qcqa$pca_noblanks
```


## Lipidomics

Number of features:

  * Total: `r lipidomics_n_qcqa[1]`
  * In 75% of a class: `r lipidomics_n_qcqa[2]`

```{r}
#| label: fig-lipidomics-heatmap
#| fig-cap: Sample-sample ICI-Kt correlations for lipid metabolites.
lipidomics_qcqa$heatmap
```

```{r}
#| label: fig-lipidomics-mediancor
#| fig-cap: Median ICI-Kt correlations for lipid metabolites.
lipidomics_qcqa$median_cor |>
		ggplot(aes(x = treatment, y = med_cor, color = treatment, shape = outlier, group = treatment)) +
		geom_sina() +
		scale_color_manual(values = color_scales$treatment) +
	labs(y = "Median Correlation")
```

```{r}
#| label: fig-lipidomics-pca
#| fig-cap: PCA plot for lipid metabolites.
lipidomics_qcqa$pca_all
```

```{r}
#| label: fig-lipidomics-pca-nopool
#| fig-cap: PCA plot for lipid metabolites, without pooled and blank samples.
lipidomics_qcqa$pca_noblanks
```

```{r}
#| label: fig-lipidomics-pca-nooutlier
#| fig-cap: PCA plot for lipid metabolites, without pooled or outlier samples.
lipidomics_qcqa$pca_nooutlier
```

## Left Censoring of Abundances

We suspect that most of the missing values are due to left censorship, or readings being below the limits of detection.
We can actually test for this using functions in {ICIKendallTau}.

P-values for the binomial test of missing features having values in each sample lower than the median in each sample are reported in @tbl-left-censorship.

```{r}
#| label: tbl-left-censorship
#| tbl-cap: Results of binomial tests for left-censorship for each of the datasets.
rna_test = broom::tidy(rna_left$test$binomial_test)
rna_test$omics = "transcriptomics"
pm_test = broom::tidy(pm_left$test$binomial_test)
pm_test$omics = "primary metabolism"
bioamines_test = broom::tidy(bioamines_left$test$binomial_test)
bioamines_test$omics = "biogenic amines"
lipidomics_test = broom::tidy(lipidomics_left$test$binomial_test)
lipidomics_test$omics = "lipidomics"
test_table = dplyr::bind_rows(rna_test, pm_test, bioamines_test, lipidomics_test)

out_table = test_table |>
	dplyr::transmute(omics = omics,
									 p.value = p.value,
									 n_under = statistic,
									 n_test = parameter,
									 under_test = estimate)
gt_table = out_table |>
	gt::gt() |>
	gt::fmt_scientific(columns = p.value)
gt_table
```

We can also plot the number of missing features vs their median rank across samples in @fig-missing-rank.

```{r}
#| label: fig-missing-rank
#| fig-cap: Median rank as a function of the number of missing values for each data set.
#| fig-width: 5
#| fig-height: 15
rna_rank = dplyr::bind_rows(rna_left$ranks$normal_adjacent$n_na_rank,
														rna_left$ranks$cancerous$n_na_rank)
rna_plot = rna_rank |>
	ggplot(aes(x = n_na, y = median_rank)) +
	geom_point(size = 0.1, position = "jitter") +
	facet_wrap(~ split, nrow = 1) +
	labs(subtitle = "RNA-Seq")

pm_rank = dplyr::bind_rows(pm_left$ranks$normal_adjacent$n_na_rank,
														pm_left$ranks$cancerous$n_na_rank)
pm_plot = pm_rank |>
	ggplot(aes(x = n_na, y = median_rank)) +
	geom_point(size = 1, position = "jitter") +
	facet_wrap(~ split, nrow = 1) +
	labs(subtitle = "Primary Metabolism")

bioamines_rank = dplyr::bind_rows(bioamines_left$ranks$normal_adjacent$n_na_rank,
														bioamines_left$ranks$cancerous$n_na_rank)
bioamines_plot = bioamines_rank |>
	ggplot(aes(x = n_na, y = median_rank)) +
	geom_point(size = 1, position = "jitter") +
	facet_wrap(~ split, nrow = 1) +
	labs(subtitle = "Biogenic Amines")

lipidomics_rank = dplyr::bind_rows(lipidomics_left$ranks$normal_adjacent$n_na_rank,
														lipidomics_left$ranks$cancerous$n_na_rank)
lipidomics_plot = lipidomics_rank |>
	ggplot(aes(x = n_na, y = median_rank)) +
	geom_point(size = 1, position = "jitter") +
	facet_wrap(~ split, nrow = 1) +
	labs(subtitle = "Biogenic Amines")

rna_plot / pm_plot / bioamines_plot / lipidomics_plot
```

## Executive Summary

* s11 and s12 (technical replicates) have really high sample-sample correlations in every -omics type, with the one oddity in primary metabolites, where one is an outlier (s11) and the other is not (s12).
* With the exception of the *primary metabolism* metabolites, everything else looks rather good.
* Heatmaps and PCA plots show clear separation of the cancerous and adjacent normal.
* Primary metabolites has two outlier samples in the adjacent normal (see @fig-pr-heatmap and @fig-pr-mediancor).
* In addition, the PCA in @fig-pr-pca-nooutlier is the only one that does not show complete separation of cancerous and adjacent normal.
* s1 is an outlier sample by median correlation in both primary metabolism and lipidomics. This *may* indicate it really is an outlier overall and should not be used for the un-paired comparison between cancer and adjacent-normal.
